src/node_url_pattern.cc

```diff
@@ -0,0 +1,705 @@
+    }
+    index++;
+  }
+  Local<Value> values[] = {
+      Array::New(isolate, inputs.data(), inputs.size()),
+      URLPatternComponentResult::ToJSObject(env, result.protocol),
+      URLPatternComponentResult::ToJSObject(env, result.username),
+      URLPatternComponentResult::ToJSObject(env, result.password),
+      URLPatternComponentResult::ToJSObject(env, result.hostname),
+      URLPatternComponentResult::ToJSObject(env, result.port),
+      URLPatternComponentResult::ToJSObject(env, result.pathname),
+      URLPatternComponentResult::ToJSObject(env, result.search),
+      URLPatternComponentResult::ToJSObject(env, result.hash),
+  };
+  DCHECK_EQ(arraysize(names), arraysize(values));
+  return Object::New(
+      isolate, Object::New(isolate), names, values, arraysize(names));
+}
+
+ada::url_pattern_options URLPattern::URLPatternOptions::FromJsObject(
+    Environment* env, Local<Object> obj) {
+  ada::url_pattern_options options{};
+  Local<Value> ignore_case;
+  if (obj->Get(env->context(), env->ignore_case_string())
+          .ToLocal(&ignore_case)) {
+    options.ignore_case = ignore_case->IsTrue();
+  }
+  return options;
+}
+
+MaybeLocal<Value> URLPattern::Hash() const {
+  auto context = env()->context();
+  return ToV8Value(context, url_pattern_.get_hash());
+}
+
+MaybeLocal<Value> URLPattern::Hostname() const {
+  auto context = env()->context();
+  return ToV8Value(context, url_pattern_.get_hostname());
+}
+
+MaybeLocal<Value> URLPattern::Password() const {
+  auto context = env()->context();
+  return ToV8Value(context, url_pattern_.get_password());
+}
+
+MaybeLocal<Value> URLPattern::Pathname() const {
+  auto context = env()->context();
+  return ToV8Value(context, url_pattern_.get_pathname());
+}
+
+MaybeLocal<Value> URLPattern::Port() const {
+  auto context = env()->context();
+  return ToV8Value(context, url_pattern_.get_port());
+}
+
+MaybeLocal<Value> URLPattern::Protocol() const {
+  auto context = env()->context();
+  return ToV8Value(context, url_pattern_.get_protocol());
+}
+
+MaybeLocal<Value> URLPattern::Search() const {
+  auto context = env()->context();
+  return ToV8Value(context, url_pattern_.get_search());
+}
+
+MaybeLocal<Value> URLPattern::Username() const {
+  auto context = env()->context();
+  return ToV8Value(context, url_pattern_.get_username());
+}
+
+bool URLPattern::HasRegExpGroups() const {
+  return url_pattern_.has_regexp_groups();
+}
+
+// Instance methods
+
+MaybeLocal<Value> URLPattern::Exec(Environment* env,
+                                   const ada::url_pattern_input& input,
+                                   std::optional<std::string_view>& baseURL) {
+  if (auto result =
+          url_pattern_.exec(input, baseURL ? &baseURL.value() : nullptr)) {
+    if (result->has_value()) {
+      return URLPatternResult::ToJSObject(env, result->value());
+    }
+    return Null(env->isolate());
+  }
+  env->ThrowTypeError("Failed to exec URLPattern");
```

Comment:
These throws (here and the "Failed to test ..." one below) need proper error codes.

Reply:
Any suggestion on error code naming?